File: src/core/help.ts
// Help rendering engine for specd.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { COMMANDS } from "./commands.js";

function getVersion(): string {
  try {
    const here = dirname(fileURLToPath(import.meta.url));
    const pkg = JSON.parse(readFileSync(join(here, "..", "..", "package.json"), "utf8")) as { version?: string };
    return pkg.version ?? "unknown";
  } catch {
    return "unknown";
  }
}

export function renderHelp(): string {
  const version = getVersion();
  const categories = {
    lifecycle: "LIFECYCLE",
    execution: "EXECUTION",
    inspection: "INSPECTION",
    meta: "META",
  };

  let out = `specd — spec-driven coding harness v${version}\n\n`;

  for (const [key, label] of Object.entries(categories)) {
    out += `${label}\n`;
    const cmds = COMMANDS.filter((c) => c.category === key);
    for (const c of cmds) {
      const cmdStr = `  ${c.usage.replace("specd ", "")}`;
      const paddedCmd = cmdStr.padEnd(30, " ");
      out += `${paddedCmd}${c.description}\n`;
    }
    out += `\n`;
  }
  out += `Use "specd help <command>" for detailed usage of a command.\n`;
  return out;
}

export function renderCommandHelp(cmdName: string): string {
  const c = COMMANDS.find((x) => x.command === cmdName);
  if (!c) {
    throw new Error(`Unknown command: ${cmdName}`);
  }

  let out = `NAME\n  specd ${c.command} — ${c.description}\n\n`;
  out += `SYNOPSIS\n  ${c.usage}\n\n`;
  out += `DESCRIPTION\n  ${c.longDescription}\n\n`;

  if (c.flags.length > 0) {
    out += `FLAGS\n`;
    for (const f of c.flags) {
      const typeStr = f.type === "string" ? " <val>" : "";
      out += `  --${f.name}${typeStr}    ${f.description}\n`;
    }
    out += `\n`;
  }

  if (c.exitCodes.length > 0) {
    out += `EXIT CODES\n`;
    for (const e of c.exitCodes) {
      out += `  ${e.code}  ${e.meaning}\n`;
    }
    out += `\n`;
  }

  if (c.examples.length > 0) {
    out += `EXAMPLE\n`;
    for (const ex of c.examples) {
      out += `  ${ex}\n`;
    }
  }

  return out;
}

export function renderHelpJson(): string {
  return JSON.stringify(COMMANDS, null, 2);
}

File: reference/internal/core/help.go
package core

import (
	"encoding/json"
	"fmt"
	"strings"
)

// Version is set at build time via -ldflags.
var Version = "dev"

// RenderHelp renders the top-level human-readable help text listing visible
// commands grouped by category.
func RenderHelp() string { return renderHelp(false) }

// RenderHelpAll renders the top-level help text like RenderHelp but also
// includes hidden commands.
func RenderHelpAll() string { return renderHelp(true) }

func renderHelp(includeHidden bool) string {
	categories := []struct {
		key   string
		label string
	}{
		{"lifecycle", "LIFECYCLE"},
		{"execution", "EXECUTION"},
		{"inspection", "INSPECTION"},
		{"meta", "META"},
	}
	var b strings.Builder
	fmt.Fprintf(&b, "specd — spec-driven coding harness %s\n\n", Version)
	for _, cat := range categories {
		fmt.Fprintf(&b, "%s\n", cat.label)
		for _, c := range Commands {
			if c.Category == cat.key && (includeHidden || !c.Hidden) {
				bare := strings.TrimPrefix(c.Usage, "specd ")
				line := "  " + bare
				for len(line) < 32 {
					line += " "
				}
				fmt.Fprintf(&b, "%s%s\n", line, c.Description)
			}
		}
		b.WriteByte('\n')
	}
	b.WriteString(`Use "specd help <command>" for detailed usage of a command.`)
	b.WriteByte('\n')
	return b.String()
}

// RenderCommandHelp renders detailed, man-page-style help (synopsis,
// description, flags, exit codes, examples) for the named command,
// returning an error if cmdName is not a known command.
func RenderCommandHelp(cmdName string) (string, error) {
	var c *CommandMeta
	for i, cmd := range Commands {
		if cmd.Command == cmdName {
			c = &Commands[i]
			break
		}
	}
	if c == nil {
		return "", fmt.Errorf("unknown command: %s", cmdName)
	}
	var b strings.Builder
	fmt.Fprintf(&b, "NAME\n  specd %s — %s\n\n", c.Command, c.Description)
	fmt.Fprintf(&b, "SYNOPSIS\n  %s\n\n", c.Usage)
	fmt.Fprintf(&b, "DESCRIPTION\n  %s\n\n", c.LongDescription)
	if len(c.Flags) > 0 {
		fmt.Fprintf(&b, "FLAGS\n")
		for _, f := range c.Flags {
			typeStr := ""
			if f.Type == "string" {
				typeStr = " <val>"
			}
			fmt.Fprintf(&b, "  --%s%s    %s\n", f.Name, typeStr, f.Description)
		}
		b.WriteByte('\n')
	}
	if len(c.ExitCodes) > 0 {
		fmt.Fprintf(&b, "EXIT CODES\n")
		for _, e := range c.ExitCodes {
			fmt.Fprintf(&b, "  %d  %s\n", e.Code, e.Meaning)
		}
		b.WriteByte('\n')
	}
	if len(c.Examples) > 0 {
		fmt.Fprintf(&b, "EXAMPLE\n")
		for _, ex := range c.Examples {
			fmt.Fprintf(&b, "  %s\n", ex)
		}
	}
	return b.String(), nil
}

// RenderHelpJSON renders the full, unfiltered command schema (including
// hidden and deprecated commands) as indented JSON.
func RenderHelpJSON() (string, error) {
	b, err := json.MarshalIndent(Commands, "", "  ")
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// RenderHelpJSONAll renders the command schema as indented JSON, including
// hidden commands only when includeHidden is true.
func RenderHelpJSONAll(includeHidden bool) (string, error) {
	commands := make([]CommandMeta, 0, len(Commands))
	for _, c := range Commands {
		if includeHidden || !c.Hidden {
			commands = append(commands, c)
		}
	}
	b, err := json.MarshalIndent(commands, "", "  ")
	if err != nil {
		return "", err
	}
	return string(b), nil
}
